### PR TITLE
ponyc: 0.42.0 -> 0.44.0, pony-corral: 0.5.3 -> 0.5.4, pony-stable: removal

### PR DESCRIPTION
**Pull request by redvers**
###### Motivation for this change

Updating the ponylang compiler to the latest version:

* Fix MacOS version mismatch warnings when linking pony programs.
* Fix calculation of "is prime" for numbers > 1321.
* Prevent non-opaque structs from being used as behaviour parameters.
* Update to LLVM 12.0.1.
* Added FileMode.u32.
* Changes to Promises Library.
* Remove zlib from libponyc as a dependency.
* Fix compile-time crash related to Pony-specific Optimizations.
* Update FilePath constructors to allow a non-partial way to create a FilePath.

###### Things done

* Updated ponyc to 0.44.0
* Updated googlebenchmark dependency to 1.5.4
* Updated googletest to 1.10.0
* Replaced a fetchurl call with a fetchFromGitHub call.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/ponyc/default.nix [hunk 1]
```diff
@@ -26,22 +26,23 @@ stdenv.mkDerivation (rec {
...
     ./disable-tests.patch
-    ./fix-libstdcpp-path.patch
     (substituteAll {
       src = ./make-safe-for-sandbox.patch;
-      googletest = fetchurl {
-        url = "https://github.com/google/googletest/archive/release-1.8.1.tar.gz";
-        sha256 = "17147961i01fl099ygxjx4asvjanwdd446nwbq9v8156h98zxwcv";
+      googletest = fetchFromGitHub {
+        owner = "google";
+        repo = "googletest";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/ponyc/default.nix:43]: Can we use the googletest from nixpkgs?

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 140189` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>pony-corral</li>
    <li>pony-stable</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ponyc</li>
  </ul>
</details>

**Comment by veprbl** [comment]
@GrahamcOfBorg build pony-corral pony-stable

File: pkgs/development/compilers/ponyc/default.nix (hunk 1 above)
Comment (redvers) [reply pkgs/development/compilers/ponyc/default.nix:43]: Great question.  I've seen upstream bump and unbump the versions for this due to issues and I'm not sure why.

I'd be hesitant to link to use the nixpkgs one unless I could pin it to a version.

Let me reach out to the developers to see if I can understand why there were issues with just tracking google and I'll get back to you.

Thank you! <3

**Comment by redvers** [comment]
> Result of `nixpkgs-review pr 140189` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
> 2 packages failed to build:
> 
>     * pony-corral
>     * pony-stable

pony-stable has been deprecated for a while (pony-corral is the replacement).
pony-corral absolutely should build.

Question - they're different entries in top-level but live in the same nixpkgs directory and should always be in sync.  Is it appropriate to remove pony-stable and update pony-corral in the same PR?

I always thought they were meant to be separate processes?

File: pkgs/development/compilers/ponyc/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/ponyc/default.nix:43]: > I'd be hesitant to link to use the nixpkgs one unless I could pin it to a version.

You can overwrite the version if you want.

**Comment by SuperSandro2000** [comment]
> Is it appropriate to remove pony-stable and update pony-corral in the same PR?

Yeah, sure.

File: pkgs/development/compilers/ponyc/default.nix (hunk 1 above)
Comment (redvers) [reply pkgs/development/compilers/ponyc/default.nix:43]: The reason for the set version is that gtest versions higher than the one they specified are broken on freebsd (which is a supported platform for the language).

It is a build-time dependency only so as long as it builds on the platforms that nix supports there's no reason why we can't use nix's version.

(Obviously - if there was a nix on freebsd that would be a different question)

I can make that change now.

**Comment by redvers** [comment]
> > Is it appropriate to remove pony-stable and update pony-corral in the same PR?
> 
> Yeah, sure.

To make sure I understand correctly - adding it to this PR is okay?

File: pkgs/development/compilers/ponyc/default.nix (hunk 1 above)
Comment (redvers) [reply pkgs/development/compilers/ponyc/default.nix:43]: (My medium-term goal is to try and get this package to the point where the r-ryantm bot will be able to do updates independent of me)

**Comment by SuperSandro2000** [comment]
> To make sure I understand correctly - adding it to this PR is okay?

Not a problem for me. It just can take a little it longer until we get this PR merged.

**Comment by redvers** [comment]
This just triggered a new release of corral.

I guess the subject of the PR would need to change to:
ponyc: 0.42.0->0.44.0, corral: 0.5.3->0.5.4, pony-stable: removed

Does that make sense?

**Comment by redvers** [comment]
Result of `nixpkgs-review pr 140189` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pony-corral</li>
    <li>ponyc</li>
  </ul>
</details>

File: pkgs/development/compilers/ponyc/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/ponyc/default.nix:43]: bsd support is not really fleshed out for nix yet. I would worry about that when it is here.

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 140189` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pony-corral</li>
    <li>ponyc</li>
  </ul>
</details>

**Review by veprbl — APPROVED** [review]

**Pull request merged**